kctf-operator/pkg/apis/kctf/v1alpha1/challenge_types.go

```diff
@@ -0,0 +1,129 @@
+// This file is responsible for generating CRD (Custom Resource Definition)
+// kubebuilder might be used to set: default values, optional fields and etc
+// +kubebuilder:validation:Required
+package v1alpha1
+
+import (
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	intstr "k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type PortSpec struct {
+	//+kubebuilder:validation:Optional
+	Name string `json:"name"`
+
+	// TargetPort is not optional
+	TargetPort intstr.IntOrString `json:"targetPort"`
+
+	//+kubebuilder:validation:Optional
+	Port int32 `json:"port"`
+
+	// Protocol is not optional
+	Protocol corev1.Protocol `json:"protocol"`
+}
+
+// Network specifications for the service
+type NetworkSpec struct {
+
+	// +kubebuilder:default:=true
+	Public bool `json:"public,omitempty"`
+
+	// +kubebuilder:default:=false
+	Dns bool `json:"dns,omitempty"`
+
+	// By default, one port is set with default values
+	// +kubebuilder:validation:Optional
+	Ports []PortSpec `json:"ports,omitempty"`
+}
+
+// Healthcheck specifications
+type HealthcheckSpec struct {
+
+	// +kubebuilder:default:=false
+	Enabled bool `json:"enabled,omitempty"`
+
+	// +kubebuilder:default:="healthcheck"
+	Container string `json:"container,omitempty"`
+}
+
+// ChallengeSpec defines the desired state of Challenge
+type ChallengeSpec struct {
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// Image used by the deployment
+	// Not optional and image should be passed by user (by now)
+	ImageTemplate string `json:"imageTemplate"`
+
+	// Shows if the challenge is ready to be deployed, if not,
+	// it sets the replicas to 0
+	// +kubebuilder:default:=false
+	Deployed bool `json:"deployed,omitempty"`
+
+	// The quantity of seconds of the proof of work
+	// +kubebuilder:default:=0
+	PowDifficultySeconds int32 `json:"powDifficultySeconds,omitempty"`
+
+	// The network specifications: if it's public or not, if it uses dns or not and specifications about ports
+	// +kubebuilder:validation:Optional
+	Network NetworkSpec `json:"network,omitempty"`
+
+	// Healthcheck checks if the challenge works
+	// If empty, healthcheck is not enabled by default
+	// +kubebuilder:validation:Optional
+	Healthcheck HealthcheckSpec `json:"healthcheck,omitempty"`
+
+	// Autoscaling features determine quantity of replicas and CPU utilization
+	// If empty, autoscaling is not enabled by default
+	// +kubebuilder:validation:Optional
+	HorizontalAutoscaling autoscalingv1.HorizontalPodAutoscalerSpec `json:"autoscaling,omitempty"`
```

Comment:
nit: rename the key to `HorizontalPodAutoscalerSpec` to match the object inside it